clients/src/main/java/org/apache/kafka/clients/consumer/internals/GroupCoordinator.java

```diff
@@ -0,0 +1,635 @@
+                future.addListener(new RequestFutureListener<Void>() {
+                    @Override
+                    public void onSuccess(Void value) {
+                        long now = time.milliseconds();
+                        heartbeat.receiveHeartbeat(now);
+                        long nextHeartbeatTime = now + heartbeat.timeToNextHeartbeat(now);
+                        client.schedule(HeartbeatTask.this, nextHeartbeatTime);
+                    }
+
+                    @Override
+                    public void onFailure(RuntimeException e) {
+                        client.schedule(HeartbeatTask.this, time.milliseconds() + retryBackoffMs);
+                    }
+                });
+            }
+        }
+    }
+
+    /**
+     * Send a request to get a new partition assignment. This is a non-blocking call which sends
+     * a JoinGroup request to the coordinator (if it is available). The returned future must
+     * be polled to see if the request completed successfully.
+     * @return A request future whose completion indicates the result of the JoinGroup request.
+     */
+    private RequestFuture<S> sendJoinGroupRequest() {
+        if (coordinatorUnknown())
+            return RequestFuture.coordinatorNotAvailable();
+
+        // send a join group request to the coordinator
+        log.debug("(Re-)joining {} group {}", groupProtocol.name(), groupId);
+
+        ByteBuffer metadata = serializeMetadata(metadata());
+        JoinGroupRequest request = new JoinGroupRequest(
+                groupProtocol.name(),
+                groupId,
+                subProtocols(),
+                this.sessionTimeoutMs,
+                this.memberId,
+                metadata);
+
+        // create the request for the coordinator
+        log.debug("Issuing request ({}: {}) to coordinator {}", ApiKeys.JOIN_GROUP, request, this.coordinator.id());
+        return client.send(coordinator, ApiKeys.JOIN_GROUP, request)
+                .compose(new JoinGroupResponseHandler());
+    }
+
+
+    private class JoinGroupResponseHandler extends CoordinatorResponseHandler<JoinGroupResponse, S> {
+
+        @Override
+        public JoinGroupResponse parse(ClientResponse response) {
+            return new JoinGroupResponse(response.responseBody());
+        }
+
+        @Override
+        public void handle(JoinGroupResponse joinResponse, RequestFuture<S> future) {
+            // process the response
+            short errorCode = joinResponse.errorCode();
+            if (errorCode == Errors.NONE.code()) {
+                log.debug("Joined group: {}", joinResponse.toStruct());
+                GroupCoordinator.this.memberId = joinResponse.memberId();
+                GroupCoordinator.this.generation = joinResponse.generationId();
+                GroupCoordinator.this.rejoinNeeded = false;
+                sensors.joinLatency.record(response.requestLatencyMs());
+                performSync(joinResponse).chain(future);
```

Comment:
Nice. It took me a minute to figure out how the combination of join group and sync were working, but the futures made this work out very nicely. This is quite a bit cleaner than I thought it was going to be.